llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp

```diff
@@ -411,29 +418,64 @@ PreservedAnalyses PrepareSYCLNativeCPUPass::run(Module &M,
   // Then we iterate over all used builtins and
   // replace them with calls to our Native CPU functions.
   for (const auto &Entry : UsedBuiltins) {
-    SmallVector<Instruction *> ToRemove;
+    SmallVector<std::pair<Instruction *, Instruction *>> ToRemove;
     Function *const Glob = Entry.first;
     for (const auto &Use : Glob->uses()) {
-      auto *ReplaceFunc = getReplaceFunc(M, Entry.second.first, StateType);
-      auto *I = cast<CallInst>(Use.getUser());
-      auto *Arg = ConstantInt::get(Type::getInt32Ty(M.getContext()),
-                                   Entry.second.second);
-      auto *NewI = CallInst::Create(
-          ReplaceFunc->getFunctionType(), ReplaceFunc,
-          {Arg, getStateArg(I->getFunction(), CurrentStatePointerTLS)},
-          "ncpu_call", I);
-      if (I->getMetadata("dbg"))
-        NewI->setDebugLoc(I->getDebugLoc());
-      I->replaceAllUsesWith(NewI);
-      ToRemove.push_back(I);
+      auto *ReplaceFunc = getReplaceFunc(M, Entry.second, StateType);
+      auto I = cast<CallInst>(Use.getUser());
+      SmallVector<Value *> Args(I->arg_begin(), I->arg_end());
+      Args.push_back(getStateArg(I->getFunction(), CurrentStatePointerTLS));
+      auto *NewI = CallInst::Create(ReplaceFunc->getFunctionType(), ReplaceFunc,
+                                    Args, "", I);
+      // If the parent function has debug info, we need to make sure that the
+      // CallInstructions in it have debug info, otherwise we end up with
+      // invalid IR after inlining.
+      if (I->getFunction()->hasMetadata("dbg")) {
+        I->setDebugLoc(DILocation::get(M.getContext(), 0, 0,
```

Comment:
Should we grab at least some of debug information from the original instruction instead of just resetting it completely?

Reply:
Yes you are right, I've changed that code to only set an empty debug location if no previous location existed, thank you